### PR TITLE
ConfStore: get_keys API

### PR DESCRIPTION
**Pull request by suryakumar1024**
Updated self._keys in kv_payload
updated existing set, delete api code to support get_keys

Signed-off-by: suryakumar <suryakumar.kumaravelan@seagate.com>

### CODE REVIEW

**Review by sachinpunadikar — CHANGES_REQUESTED** [review]
Get rid of not required print.
Also provide sample example & o/p of the same

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 1]
```diff
@@ -80,6 +80,15 @@ def delete(args):
         if any(is_deleted):
             Conf.save(ConfCli._index)
 
+    @staticmethod
+    def get_keys(args) -> str:
+        """ Returns list of keys present in store """
+        key_index_status = True if (args.key_index == None or
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_cli.py:86]:
```
key_index = 'true' if args.key_index == None else args.key_index.lower().strip()
key_index = True if key_index == 'true' else False if key_index == 'false' else None
if key_index == None:
        raise Error(...)
return Conf.get_keys(ConfCli._index, key_index=key_index)
```

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 2]
```diff
@@ -132,6 +141,22 @@ def add_args(sub_parser) -> None:
         s_parser.set_defaults(func=ConfCli.delete)
         s_parser.add_argument('args', nargs='+', default=[], help='args')
 
+class Get_KeysCmd:
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_cli.py:144]: You can remove _ from class name. Class names are with Canonical names. 

class GetKeysCommand:

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 3]
```diff
@@ -132,6 +141,22 @@ def add_args(sub_parser) -> None:
...
+    """ Get keys command structure """
+
+    @staticmethod
+    def add_args(sub_parser) -> None:
+        s_parser = sub_parser.add_parser('get_keys', help=
+            "Retrieves the list of keys\n."
+            "Example(s): ['k1', 'k1>k2','k3'], ['k4[2]>k5', 'k6>k4[2]>k5']\n\n"
+            "Example command:\n"
+            "# conf json:///tmp/csm.conf get_keys\n\n"
+            "# conf json:///tmp/csm.conf get_keys -key_index true\n\n")
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_cli.py:154]: "# conf json:///tmp/csm.conf get_keys -key_index false\n\n")

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 4]
```diff
@@ -132,6 +141,22 @@ def add_args(sub_parser) -> None:
...
+        s_parser.set_defaults(func=ConfCli.get_keys)
+        s_parser.add_argument('-key_index', dest='key_index', help=
+            "key_index={True|False} (default: True)\n"
+            "when False, returns keys including array index\n"
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_cli.py:158]: when True, returns keys including array index.

File: py-utils/src/utils/kv_store/kv_payload.py [hunk 5]
```diff
@@ -34,32 +34,45 @@ def __init__(self, data, delim='>'):
...
-
-    def _refresh_keys(self, data, pkey: str = None):
+    def get_keys(self, **filters):
+        """
+        Obtains list of keys stored in the payload
+        Input Paramters:
+        Filters - Filters to be applied before the keys to be returned.
+                  List of filters:
+                  * key_index={True|False} (default: True)
+                    when False, returns keys including array index
```
Comment (ujjwalpl) [inline py-utils/src/utils/kv_store/kv_payload.py:51]: when True, returns keys including array index

File: py-utils/src/utils/kv_store/kv_store_collection.py [hunk 6]
```diff
@@ -120,11 +120,11 @@ class IniKvPayload(KvPayload):
...
         super().__init__(configparser, delim)
 
-    def refresh_keys(self):
-        self._keys = []
+    def _get_keys(self, keys: list, data, pkey: str = None,
+        key_index: bool = True):
         for section in self._data.sections():
             for key in [option for option in self._data[section]]:
-                self._keys.append(f"{section}>{key}")
+                keys.append(f"{section}>{key}")
```
Comment (ujjwalpl) [inline py-utils/src/utils/kv_store/kv_store_collection.py:127]: keys.append(f"{section}{delim}{key}")

Please change to delim. There are few more places where '>' is used.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 1 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cli.py]: Updated in current patch

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 2 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cli.py]: Updated in current patch

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 3 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cli.py]: Updated in current patch

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 4 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cli.py]: Updated comment in current patch

File: py-utils/src/utils/kv_store/kv_payload.py (hunk 5 above)
Comment (suryakumar1024) [reply py-utils/src/utils/kv_store/kv_payload.py]: Updated comment in current patch

File: py-utils/src/utils/kv_store/kv_store_collection.py (hunk 6 above)
Comment (suryakumar1024) [reply py-utils/src/utils/kv_store/kv_store_collection.py]: Updated including inikvstore get, set and delete delim changes.

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 7]
```diff
@@ -80,6 +80,15 @@ def delete(args):
         if any(is_deleted):
             Conf.save(ConfCli._index)
 
+    @staticmethod
+    def get_keys(args) -> str:
+        """ Returns list of keys present in store """
+        key_index_status = True if (args.key_index == None or
+            args.key_index.lower().strip() == 'true') else False
+        print("*"*25)
+        print(key_index_status)
```
[the remark this answers is not in the document]
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cli.py]: Removed those 88 and 89 line statements. Updated in current patch

**Review by sachinpunadikar — APPROVED** [review]
+1

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 8]
```diff
@@ -80,6 +80,15 @@ def delete(args):
         if any(is_deleted):
             Conf.save(ConfCli._index)
 
+    @staticmethod
+    def get_keys(args) -> str:
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_cli.py:84]: This should be returning list of keys ?

File: py-utils/src/utils/conf_store/conf_store.py [hunk 9]
```diff
@@ -113,9 +113,18 @@ def set(self, index: str, key: str, val):
 
         self._cache[index].set(key, val)
 
-    def get_keys(self, index: str):
-        """ Obtains list of keys stored in the specific config store """
-        return self._cache[index].get_keys()
+    def get_keys(self, index: str, **filters):
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:116]: Returns ? i.e.g "-> list" ?

File: py-utils/src/utils/conf_store/conf_store.py [hunk 10]
```diff
@@ -198,6 +207,15 @@ def copy(src_index: str, dst_index: str, key_list: list = None):
         Conf._conf.save(dst_index)
 
     @staticmethod
-    def get_keys(index: str):
-        """ Obtains list of keys stored in the specific config store """
-        return Conf._conf.get_keys(index)
+    def get_keys(index: str, **filters):
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:210]: Returns ?

File: py-utils/src/utils/kv_store/kv_payload.py [hunk 11]
```diff
@@ -34,32 +34,45 @@ def __init__(self, data, delim='>'):
...
 
-    def get_keys(self):
-        return self._keys
-
-    def refresh_keys(self):
-        """ Refresh keys. Caller can overrite this """
-        self._refresh_keys(self._data)
-
-    def _refresh_keys(self, data, pkey: str = None):
+    def get_keys(self, **filters):
```
Comment (ujjwalpl) [inline py-utils/src/utils/kv_store/kv_payload.py:44]: Returns ?

File: py-utils/src/utils/kv_store/kv_store_collection.py [hunk 12]
```diff
@@ -120,14 +120,14 @@ class IniKvPayload(KvPayload):
     def __init__(self, configparser, delim='>'):
         super().__init__(configparser, delim)
 
-    def refresh_keys(self):
-        self._keys = []
+    def _get_keys(self, keys: list, data, pkey: str = None,
```
Comment (ujjwalpl) [inline py-utils/src/utils/kv_store/kv_store_collection.py:123]: Returns ?

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 8 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cli.py]: Updated in recent patch

File: py-utils/src/utils/conf_store/conf_store.py (hunk 9 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_store.py]: Updated in recent patch

File: py-utils/src/utils/conf_store/conf_store.py (hunk 10 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_store.py]: Updated in recent patch

File: py-utils/src/utils/kv_store/kv_payload.py (hunk 11 above)
Comment (suryakumar1024) [reply py-utils/src/utils/kv_store/kv_payload.py]: Updated in recent patch

File: py-utils/src/utils/kv_store/kv_store_collection.py (hunk 12 above)
Comment (suryakumar1024) [reply py-utils/src/utils/kv_store/kv_store_collection.py:123]: Updated in recent patch

**Review by ujjwalpl — APPROVED** [review]

**Review by s-arya — APPROVED** [review]

**Comment by mukul-seagate11** [comment]
ok to test

**Comment by mukul-seagate11** [comment]
run mini-provisioning

**Pull request merged**